packages/grafana-ui/src/components/GraphNG/GraphConfigBuilder.tsx

```diff
@@ -0,0 +1,143 @@
+import tinycolor from 'tinycolor2';
+import uPlot from 'uplot';
+import { AxisProps, ScaleProps, SeriesProps } from '../uPlot/geometries/types';
+import { calculateAxisSize, calculateSpace, formatTime } from '../uPlot/utils';
+import { PlotSeriesConfig } from '../uPlot/types';
+
+abstract class UPlotConfigBuilder<P, T> {
+  constructor(protected props: P) {}
+  abstract getConfig(): T;
+}
+
+export class GraphConfigBuilder {
+  private series: GraphSeriesBuilder[] = [];
+  private axes: GraphAxisBuilder[] = [];
+  private scales: GraphScaleBuilder[] = [];
+
+  public addAxis(props: AxisProps) {
+    this.axes.push(new GraphAxisBuilder(props));
+  }
+
+  public addSeries(props: SeriesProps) {
+    this.series.push(new GraphSeriesBuilder(props));
+  }
+
+  public addScale(props: ScaleProps) {
+    this.scales.push(new GraphScaleBuilder(props));
+  }
+
+  getConfig() {
+    const config: PlotSeriesConfig = { series: [{}] };
+    config.axes = this.axes.map(a => a.getConfig());
+    config.series = [...config.series, ...this.series.map(s => s.getConfig())];
+    config.scales = this.scales.reduce((acc, s) => {
+      return { ...acc, ...s.getConfig() };
+    }, {});
+
+    return config;
+  }
+}
+
+class GraphAxisBuilder extends UPlotConfigBuilder<AxisProps, uPlot.Axis> {
```

Comment:
would be nice to move this to a uPlot/AxisConfig file (or something like that), and have that file also contain the calculateSpace & formatTime functions in the same file, found having all that together was helpful

Reply:
agree